packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_data.dart

```diff
@@ -0,0 +1,189 @@
+  void _recalculateValues() {
+    // Set current snapshot.
+    // derived._selectedItem.value = core.selectedItem;
+
+    // Set classes to show.
+    final classes = _snapshotClasses();
+    derived.heapClasses.value = classes;
+    _updateClassStats(
+      classes: classes,
+      className: core.className,
+      singleToUpdate: derived.singleClassStats,
+      diffToUpdate: derived.diffClassStats,
+    );
+
+    // Set pathes to show.
+    final theClass =
+        derived.singleClassStats.value ?? derived.diffClassStats.value;
+    final pathEntries =
+        derived.pathEntries.value = theClass?.statsByPathEntries;
+    final pathes = theClass?.statsByPath;
+    StatsByPathEntry? byPathEntry;
+    if (core.path != null && pathes != null && pathEntries != null) {
+      final pathStats = pathes[core.path];
+      if (pathStats != null) {
+        byPathEntry = pathEntries.firstWhereOrNull((e) => e.key == core.path);
+      }
+    }
+    derived.pathEntry.value = byPathEntry;
+
+    _assertIntegrity();
+  }
+}
+
+/// Values that define what data to show on diff screen.
+///
+/// Widgets should not update the fields directly, they should use
+/// [DiffPaneController] or [DerivedData] for this.
+class CoreData {
+  /// The list contains one item that show information and all others
+  /// are snapshots.
+  ValueListenable<List<SnapshotItem>> get snapshots => _snapshots;
+  final _snapshots = ListValueNotifier(<SnapshotItem>[SnapshotDocItem()]);
+
+  /// Selected snapshot.
+  ValueListenable<int> get snapshotIndex => _snapshotIndex;
+  final _snapshotIndex = ValueNotifier<int>(0);
+
+  SnapshotItem get selectedItem => _snapshots.value[_snapshotIndex.value];
+
+  // TODO(https://github.com/flutter/devtools/issues/4539): it is unclear
+  // whether preserving the selection between snapshots should be the
+  // default behavior. Revisit after consulting with UXR.
+
+  /// Full name for the selected class (cross-snapshot).
+  HeapClassName? className;
+
+  /// Selected retaining path (cross-snapshot).
+  ClassOnlyHeapPath? path;
+
+  ValueListenable<String?> get classFilter => _classFilter;
+  final _classFilter = ValueNotifier<String?>(null);
+}
+
+/// Values that can be calculated from [CoreData] and notifiers that take signal
+/// from widgets.
+class DerivedData {
+  DerivedData(SnapshotItem selectedItem) {
+    this.selectedItem = ValueNotifier<SnapshotItem>(selectedItem);
+  }
+
+  /// Currently selected item, to take signal from the list widget.
+  late final ValueNotifier<SnapshotItem> selectedItem;
+
+  /// Classes to show.
+  final heapClasses = ValueNotifier<HeapClasses?>(null);
+
+  /// Selected single class item in snapshot, to take signal from the table widget.
+  final singleClassStats = ValueNotifier<SingleClassStats?>(null);
+
+  /// Selected diff class item in snapshot, to take signal from the table widget.
+  final diffClassStats = ValueNotifier<DiffClassStats?>(null);
```

Comment:
for the selected notifiers, can we prefix the name with selected? e.g. `selectedSingleClassStats`, `selectedDiffClassStats`, `selectedPathEntry`

this would make it more clear what the notifiers are tracking / notifying on